app/src/main/java/de/markusfisch/android/shadereditor/opengl/ShaderRenderer.java

```diff
@@ -1393,8 +1393,10 @@ private String indexTextureNames(String source) {
 
 	private static String addPreprocessorDirective(String source,
 			String directive) {
+		boolean hasVersion = source.trim().startsWith("#version");
+		directive += "#line " + (hasVersion ? 1 : 0) + "\n";
```

Comment:
This should fix the error line location by overwriting the line using a preprocessor directive.

So code like:

```glsl
void main(void) {
  gl_FragColor = vec4(1);
}
```

becomes

```glsl
#define SHADER_EDITOR 1
#line 0
void main(void) {
  gl_FragColor = vec4(1);
}
```

By setting line to 0, the next line _(which is the first visible line)_ becomes line 1.

Reply:
Unfortunately, this would inject `#line 0` twice when [SAMPLER_EXTERNAL_OES](https://github.com/markusfisch/ShaderEditor/blob/35b2974d4b0569e810854279d365a470bd959a4b/app/src/main/java/de/markusfisch/android/shadereditor/opengl/ShaderRenderer.java#L1374) is also required.

Hm, this silently injected preprocessor directives aren't too pretty anyway. Now, in the context of highlighting the error line, they make things needlessy more complex 🤔 

The only robust option I see is to compare the number of lines in the actual shader with the source in the editor and adjust the error line accordingly.

Reply:
It is true that `#line 0` will be inserted twice, but that should be intentional, because the `SAMPLER_EXTERNAL_OES` is also hidden to the user if I recall correctly (?).